### ISSUE TITLE
confused about the data leakage in honesty_control_TQA.ipynb

### ISSUE THREAD

**Issue by 336655asd** [andyzoujm/representation-engineering#4]
I am confused with the code "Contrast Vector Control" in honesty_control_TQA.ipynb. 
You use the correct answer as input along with the question to get your direction vectors. Doesn't this constitute data leakage?
I don't know if I didn't fully understand the code. Please help me to solve this problem.

```        inputs_pos_s, masks_pos_s, split_pos = prepare_decoder_only_inputs(q_batch_pos, a_batch, tokenizer, model.model.device)
        inputs_neg_s, masks_neg_s, split_neg = prepare_decoder_only_inputs(q_batch_neg, a_batch, tokenizer, model.model.device)
        split = inputs_neg_s['input_ids'].shape[1] - split_neg

        for layer_id in layer_ids:

            with torch.no_grad():
                
                _ = wrapped_model(**inputs_pos_s)
                pos_outputs = wrapped_model.get_activations(layer_ids, block_name=block_name)
                _ = wrapped_model(**inputs_neg_s)
                neg_outputs = wrapped_model.get_activations(layer_ids, block_name=block_name)
                directions[layer_id] += coeff * (pos_outputs[layer_id][:, -split:] - neg_outputs[layer_id][:, -split:]) / len(templates)

                wrapped_model.reset()
                wrapped_model.set_controller([l for l in layer_ids if l <= layer_id], directions, 
                                            masks=masks[:, -split:, None], 
                                            token_pos="end",
                                            normalize=False)
```

**Comment by andyzoujm** [comment]
Although this implementation does not support left-to-right generation, I'm not entirely sure where the data leakage could come from. 1) I'm getting the logprobs for each choice in the same way so I'm not revealing which choice is correct. 2) The model is autoregressive, so every token will only be able to attend to previous tokens so I'm not revealing future tokens to previous tokens.

Alternatively, you can think of a transformation from this implementation to a left-to-right generation format where we modify the representation before generating each new token. There's no leakage even though we performed representation control during the span of the answer string as well.

**Comment by 336655asd** [comment]
Thank you for your prompt reply, I will read your implementation carefully again. If there are new findings, I will discuss with you as soon as possible. By the way, this is indeed a valuable research.

**Comment by Atry** [comment]
The model acts as a classifier against Truthful MC1, where the answers are candidates to choose, not labels.
